File: BomPriceApproval.API/Infrastructure/Authorization/AdminOverrideAuthorization.cs
using BomPriceApproval.API.Domain.Enums;

namespace BomPriceApproval.API.Infrastructure.Authorization;

public static class AdminOverrideAuthorization
{
    private static readonly Dictionary<RequisitionStatus, RequisitionStatus> RollbackWhitelist = new()
    {
        [RequisitionStatus.Approved] = RequisitionStatus.MdReview,
        [RequisitionStatus.MdReview] = RequisitionStatus.CostingPending,
        [RequisitionStatus.CostingInProgress] = RequisitionStatus.CostingPending,
        [RequisitionStatus.CostingPending] = RequisitionStatus.BomInProgress,
        [RequisitionStatus.BomInProgress] = RequisitionStatus.BomPending,
    };

    public static bool CanRollback(RequisitionStatus from, RequisitionStatus to)
        => RollbackWhitelist.TryGetValue(from, out var allowed) && allowed == to;

    public static bool CanUnlockBom(RequisitionStatus current)
        => current is RequisitionStatus.CostingPending
                    or RequisitionStatus.CostingInProgress
                    or RequisitionStatus.MdReview;

    public static bool CanUnlockCosting(RequisitionStatus current)
        => current is RequisitionStatus.MdReview;

    /// <summary>For UI/API to populate the rollback target dropdown.</summary>
    public static RequisitionStatus? RollbackTarget(RequisitionStatus from)
        => RollbackWhitelist.TryGetValue(from, out var target) ? target : null;
}

File: bom-web/src/features/admin/adminOverrideAuthorization.ts
import type { RequisitionStatus } from "@/types/api";

const rollbackWhitelist: Partial<Record<RequisitionStatus, RequisitionStatus>> = {
  Approved: "MdReview",
  MdReview: "CostingPending",
  CostingInProgress: "CostingPending",
  CostingPending: "BomInProgress",
  BomInProgress: "BomPending",
};

export function rollbackTarget(from: RequisitionStatus): RequisitionStatus | null {
  return rollbackWhitelist[from] ?? null;
}

export function canRollback(from: RequisitionStatus): boolean {
  return rollbackTarget(from) !== null;
}

export function canUnlockBom(current: RequisitionStatus): boolean {
  return (["CostingPending", "CostingInProgress", "MdReview"] as RequisitionStatus[]).includes(current);
}

export function canUnlockCosting(current: RequisitionStatus): boolean {
  return current === "MdReview";
}

export function canDelete(): boolean {
  return true;
}

export function canReassignSp(): boolean {
  return true;
}
